content/en/blog/_posts/2021-12-07-dual-stack-networking-ga.md

```diff
@@ -0,0 +1,47 @@
+---
+layout: blog
+title: 'Kubernetes 1.23: Dual-stack IPv4/IPv6 Networking Reaches GA'
+date: 2021-12-07
+slug: dual-stack-networking-ga
+---
+
+**Author:** Bridget Kromhout (Microsoft)
+
+"When will Kubernetes have IPv6?" This question has been asked with increasing frequency ever since alpha support for IPv6 was first added in K8s 1.9. And while Kubernetes has had IPv6-only clusters since 1.18, migration from IPv4 to IPv6 was not yet possible at that point. At long last, [dual-stack IPv4/IPv6 networking](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/563-dual-stack) reaches general availability (GA) in K8s 1.23. What does dual-stack networking mean for you? Let’s take a look...
+
+
+## Service API updates
+
+Services were single-stack before 1.20, so using both IP families meant creating one Service per IP family. The user experience was simplified in 1.20, when Services were re-implemented to allow both IP families, meaning a single Service can handle both IPv4 and IPv6 workloads. Dual-stack load balancing is possible between services running any combination of IPv4 and IPv6.
+
+The Service API now has new fields to support dual-stack, replacing the single ipFamily field.
+* You can select your choice of IP family by setting `ipFamilyPolicy` to one of three options: SingleStack, PreferDualStack, or RequireDualStack. A service can be changed between single-stack and dual-stack (within some limits).
+* Setting `ipFamilies` to a list of families assigned allows you to set the order of families used.
+* `clusterIPs` is inclusive of the previous `clusterIP` but allows for multiple entries, so it’s no longer necessary to run duplicate services, one in each of the two IP families. Instead, you can assign cluster IP addresses in both IP families. Note that for a given pod, there is no possibility of setting multiple IP addresses in the same family.
+
+
+## Default behavior remains single-stack
+
+Starting in 1.20, the underlying networking for Kubernetes has included dual-stack whether or not a cluster was configured with the feature flag to enable dual-stack. Now that the feature is stable in 1.23, the feature flag is removed.
+
+However, Pods and Services default to single-stack unless `ipFamilyPolicy` specifies PreferDualStack or RequireDualStack. Even though dual-stack is possible, it is not mandatory to use it. Examples in the documentation show the variety possible in [dual-stack service configuration](/docs/concepts/services-networking/dual-stack/#dual-stack-service-configuration-scenarios).
```

Comment:
Pods default to whatever the CNI sets, this has escaped the feature gate since the beginning, I can't remember if deliberately or no.
If your CNI assigns dual stack IPs , the pod.status.PodIPs are dual stack

Reply:
Okay, so a more correct version of "However, Pods and Services default to single-stack unless `ipFamilyPolicy` specifies PreferDualStack or RequireDualStack." would be like this, perhaps...

"However, Pods and Services default to whatever the CNI sets. If your CNI assigns single-stack IPs, you will have single-stack unless `ipFamilyPolicy` specifies PreferDualStack or RequireDualStack. If your CNI assigns dual stack IPs , the pod.status.PodIPs default to dual-stack.

Reply:
I'd write “CNI plugin” or “network plugin”, not “CNI” (which is just the interface the plugin conforms to).

Reply:
Edited.